[VPlan] Remove no-op SCALAR-STEPS after unrolling.

fhahn opened the pull request:
After unrolling, there may be additional simplifications that can be applied. One example is removing SCALAR-STEPS for the first part where only the first lane is demanded.

This removes redundant adds of 0 from a large number of tests (~200), many which I am still working on updating.

In preparation for removing redundant WideIV steps added in https://github.com/llvm/llvm-project/pull/119284.

fhahn committed:
[VPlan] Simplify before covnertToDeadRecipes, remove SCALAR-STEPS.

After unrolling, there may be additional simplifications that can be
applied. One example is removing SCALAR-STEPS for the first part
where only the first lane is demanded.

This removes redundant adds of 0 from a large number of tests (~200),
many which I am still working on updating.

In preparation for removing redundant WideIV steps added in
https://github.com/llvm/llvm-project/pull/119284.

MacDue commented:
Typo in PR title, probably should be: `[VPlan] Simplify before removeDeadRecipes, remove SCALAR-STEPS.`?

fhahn commented:
> Typo in PR title, probably should be: `[VPlan] Simplify before convertToConcreteRecipes, remove SCALAR-STEPS.`?

Yes thanks! Should be fixed now

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -888,6 +888,14 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
 
+  if (auto *Steps = dyn_cast<VPScalarIVStepsRecipe>(&R)) {
+    if (Steps->getParent()->getPlan()->isUnrolled() &&
+        Steps->getNumOperands() == 2 && vputils::onlyFirstLaneUsed(Steps)) {
```
Isn't `Steps` guaranteed to have two operands? The constructor for VPScalarIVStepsRecipe seems to suggest that.

fhahn replied:
If the VPlan gets unrolled, the unrolled part will be added as extra operand. No operand implies being for the first part and we cannot simplify other parts.

I added a comment to clarify (and also added a hopefully more descriptive way to check for part 0)

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7652,6 +7652,8 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
   VPlanTransforms::unrollByUF(BestVPlan, BestUF,
                               OrigLoop->getHeader()->getContext());
   VPlanTransforms::optimizeForVFAndUF(BestVPlan, BestVF, BestUF, PSE);
+  VPlanTransforms::simplifyRecipes(BestVPlan, Legal->getWidestInductionType());
```
It looks like the code to refactor simplifyRecipes could actually be its own NFC patch, right?

fhahn replied:
Do you mean just exposing it in VPlanTransforms.h without adding new users outside of VPlanTransforms.cpp?

david-arm replied:
I just meant the changes to pull out the `removeDeadRecipes` call from `simplifyRecipes ` and call `simplifyRecipes` and `removeDeadRecipes` here from `executePlan`. It looked like these changes are notionally independent from your changes in `simplifyRecipe` for `VPScalarIVStepsRecipe`

david-arm commented:
nit: The title still says `covnertToConcreteRecipes`. I think it should be `convertToConcreteRecipes`?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-scalarivsteps-after-unroll

fhahn committed:
!fixup add clarifying comment

fhahn commented:
> nit: The title still says `covnertToConcreteRecipes`. I think it should be `convertToConcreteRecipes`?

Another typo! Should be fixed now

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-scalarivsteps-after-unroll

fhahn committed:
!fixup after merge.

fhahn committed:
[VPlan] Run recipe removal and simplification after optimizeForVFAndUF.

Run recipe simplification and dead recipe removal after VPlan-based
unrolling and optimizeForVFAndUF, to clean up any redundant or dead
recipes introduced by them. Currently this is NFC, as it removes
the corresponding removeDeadRecipes run in optimizeForVFAndUF and no
additional simplifications kick in after unrolling yet. That is changing
with https://github.com/llvm/llvm-project/pull/123655.

Note that with this change, pattern-matching is now applied after EVL-based
recipes have been introduced.

Trying to match VPWidenEVLRecipe when not explicitly requested might
apply a pattern with 2 operands to one with 3 due to the extra EVL
operand and VPWidenEVLRecipe being a subclass of VPWidenRecipe.

To prevent this, update Recipe_match::match to only match
VPWidenEVLRecipe if it is in the requested recipe types (RecipeTy).

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -890,6 +890,17 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+  // VPScalarIVSteps can only be simplified after unrolling. VPScalarIVSteps for
+  // part 0 can be replaced by their start value, if only the first lane is
+  // demanded.
+  if (auto *Steps = dyn_cast<VPScalarIVStepsRecipe>(&R)) {
```
Hi @fhahn, what I meant in one of my earlier comments was about taking everything in this patch *except* this new `VPScalarIVStepsRecipe` code. Then using that as a standalone PR, just to make it easier to see which tests are affected by this code and which are affected by the changes to `executePlan`. I tested this out downstream by building everything in this patch *except* this code and found some tests crash at Line 962 for this recipe `WIDEN ir<%mul16> = vp.mul vp<%3>, vp<%10>, vp<%7>` and the assert is:

```
bool llvm::VPlanPatternMatch::Recipe_match<Ops_t, Opcode, Commutative, RecipeTys>::match(const llvm::VPRecipeBase*) const [with Ops_t = std::tuple<llvm::VPlanPatternMatch::bind_ty<llvm::VPValue>, llvm::VPlanPatternMatch::specific_intval<0> >; unsigned int Opcode = 17; bool Commutative = true; RecipeTys = {llvm::VPWidenRecipe, llvm::VPReplicateRecipe, llvm::VPWidenCastRecipe, llvm::VPInstruction}]: Assertion `R->getNumOperands() == std::tuple_size<Ops_t>::value && "recipe with matched opcode the expected number of operands"' failed.
```

fhahn replied:
Yep I was just working on that after updating to latest main, but got sidetracked before I could put up the PR. It's here now: https://github.com/llvm/llvm-project/pull/125926

fhahn committed:
Merge branch 'vplan-simpe-cleanup-after-optimize' into vplan-simplify-scalarivsteps-after-unroll

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-scalarivsteps-after-unroll

fhahn committed:
!fixup update tests.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-scalarivsteps-after-unroll

fhahn committed:
!fixup update tests

fhahn commented:
Ping :)

All tests should be updated now

fhahn reviewed (commented):
ping :)

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-scalarivsteps-after-unroll

fhahn committed:
!fixup update new tests.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-scalarivsteps-after-unroll

fhahn reviewed (commented):
ping :)

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-simplify-scalarivsteps-after-unroll

fhahn committed:
!fixup unsigned -> bool.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -924,6 +924,17 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+  // part 0 can be replaced by their start value, if only the first lane is
+  // demanded.
+  if (auto *Steps = dyn_cast<VPScalarIVStepsRecipe>(&R)) {
+    if (Steps->getParent()->getPlan()->isUnrolled() && Steps->isPart0() &&
```
In general it is the other way around, unrolling introduces copies of the input recipe, one for each part. So only after unrolling recipes for different 'parts' (i.e. interleaved iterations) are introduced.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3593,6 +3597,9 @@ class VPlan {
   }
 
+  /// Returns true if the VPlan already has been unrolled, i.e. it has UF = 1.
+  unsigned isUnrolled() const { return UFs.size() == 1 && UFs.back() == 1; }
```
Fixed, thanks

david-arm reviewed (approved):
LGTM!

fhahn commented:
Looks like the build failures where when using GCC due to a missing explicit template instantiation. Should be fixed by b24694371c62a71aab3550e983a6bf971ed721ff
